### PR TITLE
Docker compatibility for appimage build

### PR DESCRIPTION
**Pull request by shekhar-cis**
Fixed the appimage.sh, created a dir named PyBitmessage and extracted appimagetool in it, now its working fine and appimage is in /out/ dir

### CODE REVIEW

File: buildscripts/appimage.sh [hunk 1]
```diff
@@ -4,17 +4,42 @@
 rm -rf PyBitmessage
 export VERSION=$(python setup.py --version)
 
-[ -f "pkg2appimage" ] || wget -O "pkg2appimage" https://github.com/AppImage/pkg2appimage/releases/download/continuous/pkg2appimage-1807-x86_64.AppImage
+[ -f "pkg2appimage" ] || wget -O "pkg2appimage" --no-check-certificate https://github.com/AppImage/pkg2appimage/releases/download/continuous/pkg2appimage-1807-x86_64.AppImage
```
Comment (PeterSurda) [inline buildscripts/appimage.sh:7]: `--no-check-certificate` is just for your company network.

File: buildscripts/appimage.sh [hunk 2]
```diff
@@ -4,17 +4,42 @@
...
+
 chmod a+x pkg2appimage
 
 echo "Building AppImage"
 
+
+# grep docker /proc/1/cgroup >/dev/null && export DOCKER_BUILD=1 || true
+
+if [[ ! $(cat /proc/1/sched | head -n 1 | grep init) ]]; then {
```
Comment (PeterSurda) [inline buildscripts/appimage.sh:16]:
```
unset PKG2APPIMAGE_ARGS
grep docker /proc/1/cgroup && PKG2APPIMAGE_ARGS=--appimage-extract-and-run
./pkg2appimage $PKG2APPIMAGE_ARGS /home/builder/src/packages/AppImage/PyBitmessage.yml
```

File: buildscripts/appimage.sh [hunk 3]
```diff
@@ -4,17 +4,42 @@
...
 
-[ -f "pkg2appimage" ] || wget -O "pkg2appimage" https://github.com/AppImage/pkg2appimage/releases/download/continuous/pkg2appimage-1807-x86_64.AppImage
+[ -f "pkg2appimage" ] || wget -O "pkg2appimage" --no-check-certificate https://github.com/AppImage/pkg2appimage/releases/download/continuous/pkg2appimage-1807-x86_64.AppImage
+
 chmod a+x pkg2appimage
 
 echo "Building AppImage"
 
+
+# grep docker /proc/1/cgroup >/dev/null && export DOCKER_BUILD=1 || true
```
Comment (PeterSurda) [inline buildscripts/appimage.sh:14]: remove

File: buildscripts/appimage.sh [hunk 4]
```diff
@@ -4,17 +4,42 @@
...
+
+} else {
+    echo not in docker
+    ./pkg2appimage --appimage-extract-and-run /home/builder/src/packages/AppImage/PyBitmessage.yml
+} 
+fi
+
 ./pkg2appimage packages/AppImage/PyBitmessage.yml
 
+# ./pkg2appimage /home/builder/src/packages/AppImage/PyBitmessage.yml
```
Comment (PeterSurda) [inline buildscripts/appimage.sh:28]: remove

File: buildscripts/appimage.sh [hunk 5]
```diff
@@ -4,17 +4,42 @@
...
+
+echo $(ls /)
```
Comment (PeterSurda) [inline buildscripts/appimage.sh:30]: remove

File: buildscripts/appimage.sh [hunk 6]
```diff
@@ -4,17 +4,42 @@
...
 
 if [ -f "out/PyBitmessage-${VERSION}.glibc2.15-x86_64.AppImage" ]; then
     echo "Build Successful";
     echo "Run out/PyBitmessage-${VERSION}.glibc2.15-x86_64.AppImage"
 else
     echo "Build Failed"
 fi
+
+
+# if [[ "$ARCH" = "x86_64" ]] ; then
```
Comment (PeterSurda) [inline buildscripts/appimage.sh:40]: remove

File: packages/docker/Dockerfile.bionic [hunk 7]
```diff
@@ -18,14 +18,15 @@ RUN add-apt-repository ppa:deadsnakes/ppa
...
 RUN apt-get install -yq --no-install-suggests --no-install-recommends \
     # travis xenial bionic
-    python-setuptools libssl-dev libpq-dev python-prctl python-dev \
+    python-setuptools libssl-dev libpq-dev python-prctl \
     python-dev python-virtualenv python-pip virtualenv \
     # dpkg
-    python-minimal python-setuptools python-all python openssl libssl-dev \
-    dh-apparmor debhelper dh-python python-msgpack python-qt4 python-stdeb \
-    python-all-dev python-crypto python-psutil \
+    libfuse2 libglib2.0-bin python-minimal python-all python openssl libssl-dev \
```
Comment (PeterSurda) [inline packages/docker/Dockerfile.bionic:27]: don't add `libfuse2`

File: packages/docker/Dockerfile.bionic [hunk 8]
```diff
@@ -18,14 +18,15 @@ RUN add-apt-repository ppa:deadsnakes/ppa
...
+    dh-apparmor debhelper dh-python python-msgpack python-qt4 git python-stdeb \
+    python-all-dev python-crypto python-psutil python-pycryptopp \
```
Comment (PeterSurda) [inline packages/docker/Dockerfile.bionic:29]: I don't think you need `python-pycryptopp`

File: packages/docker/Dockerfile.bionic [hunk 9]
```diff
@@ -84,4 +85,29 @@ RUN echo 'buildbot ALL=(ALL) NOPASSWD:ALL' >> /etc/sudoers
...
 
-ENTRYPOINT /entrypoint.sh "$BUILDMASTER" "$WORKERNAME" "$WORKERPASS"
+ENTRYPOINT /entrypoint.sh $BUILDMASTER $WORKERNAME" "$WORKERPASS"
+
+#####################################################################################################
+
+FROM base AS appimage
+
+# travis2bash
+RUN wget -O /usr/local/bin/travis2bash.sh https://git.bitmessage.org/Bitmessage/buildbot-scripts/raw/branch/master/travis2bash.sh
```
Comment (PeterSurda) [inline packages/docker/Dockerfile.bionic:95]: we don't need travis2bash for building an appimage

File: packages/docker/Dockerfile.bionic [hunk 10]
```diff
@@ -84,4 +85,29 @@ RUN echo 'buildbot ALL=(ALL) NOPASSWD:ALL' >> /etc/sudoers
...
+
+RUN python setup.py sdist
+
+RUN python setup.py  --command-packages=stdeb.command  bdist_deb
+
+RUN dpkg-deb -I deb_dist/pybitmessage_0.6.3.2-1_amd64.deb
+
+RUN buildscripts/appimage.sh
+
+RUN out/PyBitmessage-0.6.3.2.glibc2.15-x86_64.AppImage
```
Comment (PeterSurda) [inline packages/docker/Dockerfile.bionic:112]: since we're running inside docker, we need to add `--appimage-extract-and-run` when running an appimage.

**Review by PeterSurda — CHANGES_REQUESTED** [review]

File: buildscripts/appimage.sh [hunk 11]
```diff
@@ -9,12 +9,18 @@ chmod a+x pkg2appimage
...
 
+./pkg2appimage packages/AppImage/PyBitmessage.yml
 
 if [ -f "out/PyBitmessage-${VERSION}.glibc2.15-x86_64.AppImage" ]; then
     echo "Build Successful";
     echo "Run out/PyBitmessage-${VERSION}.glibc2.15-x86_64.AppImage"
 else
     echo "Build Failed"
-fi
+fi
```
Comment (PeterSurda) [inline buildscripts/appimage.sh:26]: newline at the end

File: packages/docker/Dockerfile.bionic [hunk 12]
```diff
@@ -18,15 +18,16 @@ RUN add-apt-repository ppa:deadsnakes/ppa
 
 RUN apt-get -y install sudo
 
+
```
Comment (PeterSurda) [inline packages/docker/Dockerfile.bionic:21]: extra line

File: packages/docker/Dockerfile.bionic [hunk 13]
```diff
@@ -84,4 +85,24 @@ RUN echo 'buildbot ALL=(ALL) NOPASSWD:ALL' >> /etc/sudoers
...
+RUN python setup.py sdist
+
+RUN python setup.py  --command-packages=stdeb.command  bdist_deb
+
+RUN dpkg-deb -I deb_dist/pybitmessage_0.6.3.2-1_amd64.deb
+
+RUN buildscripts/appimage.sh
+
+RUN out/PyBitmessage-0.6.3.2.glibc2.15-x86_64.AppImage --appimage-extract-and-run -t
```
Comment (PeterSurda) [inline packages/docker/Dockerfile.bionic:108]: newline at the end

File: packages/docker/Dockerfile.bionic [hunk 14]
```diff
@@ -84,4 +85,24 @@ RUN echo 'buildbot ALL=(ALL) NOPASSWD:ALL' >> /etc/sudoers
...
+
+FROM base AS appimage
+
+COPY . /home/builder/src
+
+WORKDIR /home/builder/src
+
+RUN python setup.py -V
+
+RUN python setup.py sdist
```
Comment (PeterSurda) [inline packages/docker/Dockerfile.bionic:100]: I would put the dpkg lines together, and the appimage lines also together. Like
```
RUN sdist
RUN bdist
RUN dpkg-deb

RUN app image
RUN out/appimage -t
```

**Review by PeterSurda — CHANGES_REQUESTED** [review]

File: packages/docker/Dockerfile.bionic (hunk 14 above)
Comment (shekhar-cis) [reply packages/docker/Dockerfile.bionic:100]:
```
RUN python setup.py sdist
RUN python setup.py --command-packages=stdeb.command bdist_deb
RUN dpkg-deb -I deb_dist/pybitmessage_0.6.3.2-1_amd64.deb

RUN buildscripts/appimage.sh
RUN out/PyBitmessage-0.6.3.2.glibc2.15-x86_64.AppImage --appimage-extract-and-run -t
```
Like this?

Comment (PeterSurda) [reply packages/docker/Dockerfile.bionic:100]: yes

File: packages/docker/Dockerfile.bionic [hunk 15]
```diff
@@ -84,4 +84,21 @@ RUN echo 'buildbot ALL=(ALL) NOPASSWD:ALL' >> /etc/sudoers
 
 USER buildbot
 
-ENTRYPOINT /entrypoint.sh "$BUILDMASTER" "$WORKERNAME" "$WORKERPASS"
+ENTRYPOINT /entrypoint.sh $BUILDMASTER "$WORKERNAME" "$WORKERPASS"
```
Comment (PeterSurda) [inline packages/docker/Dockerfile.bionic:87]: quotes missing around `$BUILDMASTER`

**Review by PeterSurda — APPROVED** [review]

**Pull request merged**